File: Optymization/algorithms/tabu_search.py
from algorithms.optimizer import Optimizer
from copy import deepcopy
from sys import maxsize


class TabuSearch(Optimizer):

    def __init__(self, nb_proc: int, data: list, max_it: int, tabu_size: int):
        super().__init__(nb_proc, data, max_it)
        self.__tabu_size = tabu_size
        self.__tabus = []

    def search(self, rng: bool) -> dict:
        """
        :return:
        """
        # s0
        current_sol = self.gen_s0(rng)
        # s*
        best_sol    = deepcopy(current_sol)
        # f(s*)
        best_time = self.c_max(best_sol)

        best_neighbor = {}

        while not self.stop_search():
            # neighbors handing
            best_neighbor_time = maxsize
            for n in self.get_neighbors(current_sol):
                if n in self.__tabus:
                    continue

                local_time = self.c_max(n)
                if local_time <= best_neighbor_time:
                    best_neighbor = n
                    best_neighbor_time = local_time

            # setting best neighbor ar current solution
            current_sol = best_neighbor

            # checking if the solution is faster
            local_time = self.c_max(current_sol)
            if local_time <= best_time:
                print('choosen: {} <= {}'.format(local_time, best_time))
                print_dict(current_sol)
                print('-------------------')
                best_sol  = deepcopy(current_sol)
                best_time = local_time

            # adding to tabus
            self.__tabus.append(deepcopy(current_sol))

            # reducing tabus
            if len(self.__tabus) > self.__tabu_size:
                self.__tabus.remove(
                    self.__tabus[::-1][0]
                )

        yield best_sol
        # return best_sol

    def stop_search(self) -> bool:
        """
        :return:
        """
        self._iteration += 1
        return self._max_it < self._iteration


def print_dict(d):
    for key, values in d.items():
        print(
            '**\t' + str(key) + ': ' + str(values)
            + '\n**\t\ttotal: ' + str(sum(values))
        )

if __name__ == '__main__':
    solver = TabuSearch(
        3,
        [5, 7, 6, 2, 3, 1],
        100,
        100
    )

    i = 0
    for sol in solver.search(True):
        # print('\n--------------------- ' + str(i))
        # print('** CMax: ' + str(Optimizer.c_max(sol)))
        # for key, values in sol.items():
        #     print(
        #         '**\t' + str(key) + ': ' + str(values)
        #         + '\n**\t\ttotal: ' + str(sum(values))
        #     )
        # i += 1
        pass

File: src/optimization/algorithms/TabuSearch.java
package optimization.algorithms;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static cli.Text.ALGO_TABU;

public class TabuSearch extends OptimizationAlgorithm {
    /**
     * Default constructor
     *
     * @param nbProcess number of process among which split tasks
     */
    public TabuSearch(int nbProcess) {
        super(nbProcess) ;
    }

    /**
     * Check is this neighbor already exists
     * @param visited list of key for visited states
     * @param node state to check
     * @return true if already visited, false otherwise
     */
    private boolean notVisited(ArrayList<String> visited, HashMap<Integer, int[]> node) {
        String currentKey = getKeyFor(node) ;

        for (String pastNodeKey : visited) {
            if (pastNodeKey.equals(currentKey)) return false ;
        }

        return true ;
    }

    /**
     * Evaluate the tabu search for current data
     *
     * @param durations list of task durations
     * @return best repartition found
     */
    @Override
    public HashMap<Integer, int[]> searchFor(List<Integer> durations, HashMap<Integer, int[]> state) {
        // keeping track of the last solution time
        int currentSolutionTime ;

        // generating s0
        currentSolution = state ;

        // s* = s0 -> first solution is the current best one
        currentBestSolution = new HashMap<>(currentSolution) ;

        // f(s*) -> current best time
        bestSolutionTime = c_max(currentSolution) ;

        // initializing neighbor handling
        ArrayList<HashMap<Integer, int[]>> neighborList ;
        HashMap<Integer, int[]> bestNeighbor ;

        int neighborBestLocalTime ; // f(x) for best neighbor
        int neighborLocalTime ;     // f(x) for current neighbor

        // getting next neighbor
        neighborList = getNeighbors() ;

        // initializing neighbor search with the first one
        bestNeighbor = neighborList.get(0) ;
        neighborBestLocalTime = c_max (bestNeighbor) ;

        // for each neighbor
        for (HashMap<Integer, int[]> neighbor : neighborList) {
            // ignoring the first one (already initialized)
            if (bestNeighbor == neighbor) continue ;

            // ignoring neighbor if it is already visited
            if (!notVisited(visited, neighbor)) continue ;

            // if f(x) doesn't improve the solution, ignoring
            if ((neighborLocalTime = c_max(neighbor))
                    > neighborBestLocalTime) continue ;

            // updating best neighbor and best neighbor's time
            bestNeighbor = neighbor ;
            neighborBestLocalTime = neighborLocalTime ;
        }

        // setting best neighbor as current solution
        currentSolution = bestNeighbor ;

        // if this solution is faster
        if ((currentSolutionTime = c_max(currentSolution)) <= bestSolutionTime) {
            if (currentSolutionTime < bestSolutionTime) stagnation = 0 ;
            else ++stagnation ;

            // s* = s
            currentBestSolution = new HashMap<>(currentSolution) ;
            // updating f(s*)
            bestSolutionTime = currentSolutionTime ;
        } else {
            ++stagnation ;
        }

        // updating tabus -> adding current state to visited
        visited.add (getKeyFor (currentBestSolution)) ;

        // checking tabu size
        if (visited.size() == tabuSize) visited.remove(0) ;

        return new HashMap<>(currentBestSolution) ;
    }

    @Override
    public int getID() {
        return ALGO_TABU ;
    }

    @Override
    public boolean stopSearch() {
        return stagnation > stagnationTrigger;
    }

    @Override
    public String toString() {
        return "Tabu search" ;
    }
}
